### ISSUE TITLE
Explore using __slots__ on Seq, SeqRecord, etc

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#2854]
Defining ``__slots__`` on the core sequence object would be a useful optimisation, but does potentially have downsides if it stops people adding their own attributes at run time. See:

https://twitter.com/raymondh/status/1255228879763054598

https://twitter.com/raymondh/status/1255246344551309313

It would be particularly interesting to see if this make a noticeable difference to e.g. parsing large files with SeqIO.

**Comment by peterjc** [comment]
Defining slots on the Seq object would help with enforcing the expectation that it is immutable (like a string).

**Comment by trkohler** [comment]
Hi! 
I'd like to get familiar with slots concept more. I read it's like an immutable version of the dictionary where obj properties contained, isn't it?
If you don't mind, I'd get this issue 🔧

**Comment by peterjc** [comment]
If you have the time it would be an interesting experiment to make a new branch where you add slots to the ``Seq`` class in ``Bio/Seq.py``, and see if our test suite still passes, and if it makes any difference to some simple benchmarks like parsing a large FASTA files (millions of sequences as in high throughput sequencing).

**Comment by trkohler** [comment]
Can I somehow run tests only related to the Seq class? I have a weak machine and running a full test suite (even offline) takes a long time.

Also, silly question, how can I check the difference? Is there anything in the project I can use for doing this? I'd be happy for the help, I'm still not familiar with a lot of things.

And, finally, am I right - Seq obj has 2 properties: data and alphabet?

**Comment by chris-rands** [comment]
You can run any of the .py files under the `Test/` folder individually. For benchmarking timing, you could modify this script  https://github.com/biopython/biopython/pull/1808#issuecomment-433043873 - benchmarking memory (where the gains of `__slots__` might be greatest) is a little tricker.

Also if there is still the plan to drop alphabets this year(?), then the Seq object might not be the best place to start using `__slots__`. Without alphabets, Seq could subclass str directly

**Comment by mdehoon** [comment]
> Also if there is still the plan to drop alphabets this year(?), 

Yes there is. For the upcoming 1.77 release, we focused on the python2 -> python3 transition, removing all python2-specific code. After 1.77 is out, we can start removing the alphabets.

> then the Seq object might not be the best place to start using __slots__. Without alphabets, Seq could subclass str directly

Python offers a better solution nowadays, which is to let Seq implement the buffer protocol. It will still effectively look like a string, but it has the advantage that you won't have to create, read in, and keep in memory the whole sequence. This would allow you to open 2bit files and effectively get random access to the whole genome without having to read it in.

Perhaps this is also a good time to consider whether we need SeqRecords at all; we could just use Seq objects instead. It's confusing and error-prone to have both.

**Comment by MarkusPiotrowski** [comment]
> Perhaps this is also a good time to consider whether we need SeqRecords at all; we could just use Seq objects instead. It's confusing and error-prone to have both.

@mdehoon I don't understand this. Isn't a SeqRecord object something that represents a sequence database entry and does contain more than just the sequence (e.g. ID, name, description, features, annotations)? Where do you want to put this information?

**Comment by mdehoon** [comment]
@MarkusPiotrowski  
> Isn't a SeqRecord object something that represents a sequence database entry and does contain more than just the sequence (e.g. ID, name, description, features, annotations)?

Yes.

> Where do you want to put this information?

As attributes of the `Seq` object. So instead of having attributes on a `SeqRecord` object, we put them on the `Seq` object directly.

**Comment by chris-rands** [comment]
> Python offers a better solution nowadays, which is to let Seq implement the buffer protocol.

@mdehoon is that similar to how Biotite stores sequence objects using NumPy ndarray objects? e.g. https://github.com/biotite-dev/biotite/blob/master/src/biotite/sequence/sequence.py#L25

> Perhaps this is also a good time to consider whether we need SeqRecords at all; we could just use Seq objects instead. It's confusing and error-prone to have both.

How to handle backward compatibility?

**Comment by MarkusPiotrowski** [comment]
> > Where do you want to put this information?
> 
> As attributes of the `Seq` object. So instead of having attributes on a `SeqRecord` object, we put them on the `Seq` object directly.

Mmh, didn't we remove attributes from the Seq object (like the sequence type) with the aim to subclass Seq objects directly from `str`?

**Comment by peterjc** [comment]
I do see attractions in making ``Seq`` subclass string (or use the buffer protocol), but like Markus don't see don't see the rational to merge the Seq and SeqRecord.

Anyway, Python strings are immutable, so you cannot add attributes at run time. Adding ``__slots__`` to the current ``Seq`` object would help make it more like the immutable string. i.e. This should be a useful experiment for some of what Michael has in mind.

```python
>>> from Bio.Seq import Seq
>>> s = Seq("ACGT")
>>> s.extra = 1234
```

versus:

```python
>>> s = "ACGT"
>>> s.extra = 1234
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'str' object has no attribute 'extra'
```

**Comment by mdehoon** [comment]
@chris-rands 
> Python offers a better solution nowadays, which is to let Seq implement the buffer protocol.
> @mdehoon is that similar to how Biotite stores sequence objects using NumPy ndarray objects? 

It's different. Biotite still stores the sequence information as an attribute to their sequence object.

**Comment by mdehoon** [comment]
@MarkusPiotrowski 
>      Mmh, didn't we remove attributes from the Seq object (like the sequence type) with the aim to subclass Seq objects directly from str?

I think these issues are unrelated. You can subclass Seq objects directly from str and still have attributes:
```
>>> class Seq(str):
...     pass
... 
>>> s = Seq("ACGT")
>>> s.id = 'mysequence'
>>> s.id
'mysequence'
>>> s.name = 'sequence1'
>>> s.name
'sequence1'
>>> 'AC' in s
True
```

**Comment by mdehoon** [comment]
@peterjc 
> Anyway, Python strings are immutable, so you cannot add attributes at run time. 

You can let `Seq` subclass from `str`, and add attributes to them, and from Python's viewpoint they are act as immutable objects:

```
>>> class Seq(str):
...     pass
... 
>>> s = Seq("ABCD")
>>> s.name = 'sequence1'
>>> d = {}
>>> d[s] = 'hello'
>>> d
{'ABCD': 'hello'}
```

**Comment by mdehoon** [comment]
@peterjc 
> I do see attractions in making Seq subclass string (or use the buffer protocol), but like Markus don't see don't see the rational to merge the Seq and SeqRecord.

We should ask: What is the rationale for having both a `Seq` and a `SeqRecord`? Other than historical reasons...

**Comment by peterjc** [comment]
Is your answer related to ``__slots__`` because we're going way off topic here?

**Comment by mdehoon** [comment]
> Is your answer related to __slots__ because we're going way off topic here?

Yes. I don't see which problem adding `__slots__` will solve, and if I am right that `Seq` and `SeqRecord` can be merged, then adding `__slots__` is a step in the wrong direction.

**Comment by trkohler** [comment]
I have some issues and I don't understand if there is my lack of knowledge or __ slots__ on seq just  can't be implemented 

I added __slots__ on Seq class:
`__slots__ = ('_data', 'alphabet')`

and I have almost all tests failed.
like not just one or two test cases, but around 80%.

**Comment by peterjc** [comment]
As per the links at the start of this issue, adding ``__slots_`` can be a useful optimisation technique. A major part of the slowness of parsing files with millions of sequences is the object creation overhead, so potentially adding slots to key objects like the ``Seq`` and ``SeqRecord`` objects could speed this up.

However, it would be a backward incompatible change, and assuming @trkohler's initial results stand, too big a change to consider.

**Comment by peterjc** [comment]
I've tried adding slots as per @trkohler's code snippet and it has the expected result:

```
>>> from Bio.Seq import Seq
>>> s = Seq("ACGT")
>>> s.extra = 1234
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Seq' object has no attribute 'extra'
```

I have run the test suite locally (offline) but saw no failures...

```
$ python setup.py install .
$ cd Tests/
$ python run_tests.py --offline
Skipping any tests requiring internet access
Python version: 3.7.6 | packaged by conda-forge | (default, Jan  7 2020, 22:05:27) 
[Clang 9.0.1 ]
Operating system: posix darwin
test_Ace ... ok
...
BioSQL.Loader docstring test ... ok
----------------------------------------------------------------------
Ran 525 tests in 178.502 seconds
```

(There were some warnings, including needing to update the BLAST+ wrappers, but that is an aside)

@trkohler I suspect something else is wrong in your local tests. It might be worth making a pull request to run the slots changes on TravisCI and AppVeyor, covering Linux & Windows with a full range of Python versions.

If that works, the real question remains - does this show any performance improvement (which might justify the behaviour change).

**Comment by mdehoon** [comment]
> As per the links at the start of this issue, adding __slots_ can be a useful optimisation technique. A major part of the slowness of parsing files with millions of sequences is the object creation overhead, so potentially adding slots to key objects like the Seq and SeqRecord objects could speed this up.

I wonder how much speedup you would get if you only had to create one object instead of two.

**Comment by trkohler** [comment]
`ImportError: cannot import name '_aligners' from partially initialized module 'Bio.Align' (most likely due to a circular import) (/Users/Wnb/Desktop/biopython/Bio/Align/__init__.py)`

I understand I can't run the script mentioned by @chris-rands in the root folder. I tried python console and no effect also. How can I solve this?

**Comment by peterjc** [comment]
@trkohler that error suggests you have not compiled the C code (for the ``_aligners`` module), and/or are not running the test script as expected.

**Comment by kaskales** [comment]
@trkohler @peterjc Could I try implementing `__slots__ ` for the Seq and SeqRecord classes?

**Comment by peterjc** [comment]
@kaskales You are welcome to try, and submit a pull request to see if it breaks anything in the test suite. The nature of this is exploration - there is less chance of the PR being accepted than usual.

The next step would be benchmarking to see if it has a noticeable effect on the speed for parsing large FASTA or FASTQ files (and perhaps memory usage), and if so we can discuss if it is worth adopting this (despite blocking adding attributes at run time).

If there is no noticeable performance improvement, then there is no reason to make the change.

**Comment by peterjc** [comment]
With large files (millions of sequences), @kaskales is reporting a significant memory saving on #3309 - enough to justify a closer look.

@mdehoon you've been benchmarking other sequence related optimisations, would you be able to try that branch to give some independent measurements?

**Comment by mdehoon** [comment]
@peterjc  I am still in favor of unifying `Seq` & `SeqRecord` into a single class.

**Comment by peterjc** [comment]
@mdehoon So you have said (I am not convinced). But what do you think about using ``__slots__`` for better performance at the cost of preventing adding attributes at run time?

**Comment by mdehoon** [comment]
How do you plan to implement the deprecation policy for this case?

**Comment by peterjc** [comment]
Adding an in-code warning seems possible via a ``__setattr__`` method:

```
diff --git a/Bio/Seq.py b/Bio/Seq.py
index 741bc64ff..597d80758 100644
--- a/Bio/Seq.py
+++ b/Bio/Seq.py
@@ -73,6 +73,14 @@ class Seq:
     not applicable to protein sequences).
     """
 
+    def __setattr__(self, name, value):
+        if name != "_data":
+            warnings.warn(
+                "Custom Seq object attributes will be blocked in a future Biopython release.",
+                BiopythonDeprecationWarning,
+            )
+        object.__setattr__(self, name, value)
+
     def __init__(self, data):
         """Create a Seq object.
```

**Comment by mdehoon** [comment]
But then we are not offering any alternative to a user who was setting additional attributes. For a `SeqRecord` a user could store such attributes in `record.annotations` for example, but for `Seq` there is no workaround.

**Comment by peterjc** [comment]
I've closed #3309 on the grounds the performance impact was minimal.

I think this issue has served its purpose, see #3482 about ``__slots__`` on the sequence objects.

Thanks all.
